src/mcpwm.rs

```diff
@@ -0,0 +1,597 @@
+                $timer
+            }
+
+            fn signal_a() -> mcpwm_io_signals_t {
+                $signal_a
+            }
+
+            fn signal_b() -> mcpwm_io_signals_t {
+                $signal_b
+            }
+        }
+    };
+}
+
+macro_rules! impl_operator {
+    ($instance:ident: $timer:expr, $signal_a:expr, $signal_b:expr) => {
+        pub struct $instance<U: Unit> {
+            _unit: U,
+        }
+
+        impl<U: Unit> $instance<U> {
+            /// # Safety
+            ///
+            /// It is safe to instantiate this operator exactly one time per Unit.
+            pub unsafe fn new() -> Self {
+                $instance {
+                    _unit: U::default(),
+                }
+            }
+        }
+
+        impl_operator_helper!($instance: $timer, $signal_a, $signal_b, UnitZero);
+        impl_operator_helper!($instance: $timer, $signal_a, $signal_b, UnitOne);
+    };
+}
+
+impl_operator!(
+    OPERATOR0: mcpwm_timer_t_MCPWM_TIMER_0,
+    mcpwm_io_signals_t_MCPWM0A,
+    mcpwm_io_signals_t_MCPWM0B
+);
+impl_operator!(
+    OPERATOR1: mcpwm_timer_t_MCPWM_TIMER_1,
+    mcpwm_io_signals_t_MCPWM1A,
+    mcpwm_io_signals_t_MCPWM1B
+);
+impl_operator!(
+    OPERATOR2: mcpwm_timer_t_MCPWM_TIMER_2,
+    mcpwm_io_signals_t_MCPWM2A,
+    mcpwm_io_signals_t_MCPWM2B
+);
+
+// TODO: How do we want syncing to fit in to this?
+// TODO: How do we want deadtime to fit into this?
+// TODO: How do we want carrier to fit into this?
+// TODO: How do we want capture to fit into this?
+
+pub struct Operator<U: Unit, O: HwOperator<U>, M: Borrow<Mcpwm<U>>, PA: OutputPin, PB: OutputPin> {
+    _instance: O,
+    _unit: U,
+    _mcpwm_module: M,
+
+    _pin_a: Option<PA>,
+    _pin_b: Option<PB>,
+}
+
+impl<U: Unit, O: HwOperator<U>, M: Borrow<Mcpwm<U>>, PA: OutputPin, PB: OutputPin>
+    Operator<U, O, M, PA, PB>
+{
+    pub fn new<A: Into<Option<PA>>, B: Into<Option<PB>>>(
+        operator: O,
+        mcpwm_module: M,
+        config: &OperatorConfig,
+        pin_a: A,
+        pin_b: B,
+    ) -> Result<Self, EspError> {
+        #[cfg(not(esp_idf_version = "4.3"))]
+        {
+            if config.frequency < config.lowest_frequency {
+                // Can not specify a clock frequency lower then what has
+                // been configured as the lowest clock frequency
+                // Use `OperatorConfig::lowest_frequency` to enable lower frequencies
+                return Err(EspError::from(ESP_ERR_INVALID_ARG).unwrap());
+            }
+
+            if config.lowest_frequency > mcpwm_module.borrow().operator_input_frequency.Hz() {
+                // Can not specify a lowest_frequency larger than the corresponding value for
+                // the parent MCPWM module. Use `Mcpwm::lowest_frequency` to enable higher frequencies
+                return Err(EspError::from(ESP_ERR_INVALID_ARG).unwrap());
+            }
+
+            let resolution = u32::from(config.lowest_frequency) * MAX_PWM_TIMER_PERIOD;
```

Comment:
This is related to what I [mentioned earlier](https://github.com/esp-rs/esp-idf-hal/pull/93/files#r926141725)

Reply:
Should we clamp `resolution` to not be greater than `mcpwm_module.operator_source_frequency`?